File: 2023/bin/day01.ml
open Core

let data = Advent.read_file "input/day01.txt"

let rec get_first_digit char_list =
  match char_list with
  | [] -> ""
  | ch :: lst -> if Char.is_digit ch then Char.escaped ch else get_first_digit lst
;;

let get_first_last_digit str =
  let ch_lst = String.to_list str in
  let first_digit = get_first_digit ch_lst in
  let last_digit = get_first_digit (List.rev ch_lst) in
  let first_last_digit = first_digit ^ last_digit in
  int_of_string first_last_digit
;;

let rec sum = function
  | [] -> 0
  | h :: t -> h + sum t
;;

let () =
  let numbers = List.map data ~f:get_first_last_digit in
  printf "Part 1: %d\n" (sum numbers)
;;

let rec _replace_cases cases str =
  match cases with
  | [] -> str
  | (case, repl) :: rest ->
    String.substr_replace_all ~pattern:case ~with_:repl (_replace_cases rest str)
;;

let replace_cases str =
  let cases_replacements =
    [ "one", "one1one"
    ; "two", "two2two"
    ; "three", "three3three"
    ; "four", "four4four"
    ; "five", "five5five"
    ; "six", "six6six"
    ; "seven", "seven7seven"
    ; "eight", "eight8eight"
    ; "nine", "nine9nine"
    ]
  in
  _replace_cases cases_replacements str
;;

let () =
  let numbers =
    List.map data ~f:(fun line -> replace_cases line |> get_first_last_digit)
  in
  printf "Part 2: %d\n" (sum numbers)
;;

File: 2020/src/day01.js
const fs = require("fs");

const data = fs
  .readFileSync("input/day01.txt")
  .toString()
  .split("\n")
  .filter(Boolean)
  .map(Number);
const result_sum = 2020;
const do_part2 = true;

for (let i = 0; i < data.length; i++) {
  const x1 = data[i];
  for (let j = i + 1; j < data.length; j++) {
    const x2 = data[j];
    // Part 1
    x1 + x2 === result_sum && console.log("Part 1: %d", x1 * x2);
    if (do_part2) {
      for (let k = j + 1; k < data.length; k++) {
        const x3 = data[k];
        // Part 2
        x1 + x2 + x3 === result_sum && console.log("Part 2: %d", x1 * x2 * x3);
      }
    }
  }
}
